Document provider_mappings database structure (#4466)
music-assistant/server: AGENTS.md

- **Logs:** `$HOME/.musicassistant/musicassistant.log` (current), `musicassistant.log.1`, `.log.2`, etc. for older rotated logs.
- **Database:** `$HOME/.musicassistant/library.db` — query via `sqlite3`. **Only execute SELECT queries** — never write to a live database.

### Database Structure

> #### `provider_mappings` table

Reviewer: this is not only a table but an attribute on every MediaItem.
I would make this less about the database structure but more about the model structure

Author: Fixed in 53711808. Rewrote the section to focus on the MediaItem.provider_mappings attribute structure rather than the database table. Now describes the model/object perspective with the correct field meanings (item_id = external provider ID, not library DB ID) and the proper resolution pattern via get_library_item_by_prov_id().

Critical structure for resolving tracks between library and external providers:

- `item_id` (integer): **ALWAYS the library database ID** — this is the key for lookups
- `provider_domain` (string): The external provider identifier (e.g., `"spotify"`, `"apple_music"`, `"tidal"`)
- `provider_item_id` (string): The external provider's ID for this item

> - `media_type` (string): Type of media item (`"track"`, `"album"`, `"artist"`, `"playlist"`)

Author: Thanks for the review. This documentation is intentionally focused on the most common mistakes that lead to bugs (specifically the non-existent `provider_domain='library'` pattern). 

The goal is to provide quick guidance for AI-assisted code generation, not a complete schema reference. Adding all columns would dilute the key message and make it harder to scan.

A full schema reference would be better suited for dedicated database documentation.

- **Logs:** `$HOME/.musicassistant/musicassistant.log` (current), `musicassistant.log.1`, `.log.2`, etc. for older rotated logs.
- **Database:** `$HOME/.musicassistant/library.db` — query via `sqlite3`. **Only execute SELECT queries** — never write to a live database.

### Provider Mappings

> Every `MediaItem` (track, album, artist, playlist) has a `provider_mappings` attribute containing mappings between library items and external providers:

Reviewer: ```suggestion
Every `MediaItem` (track, album, artist, playlist) has a `provider_mappings` attribute containing the exact mappings of the item on (each) provider. For a MediaItem that comes from a musicprovider directly, this will usually contain one single ID-mapping (although it is possible that a provider has multiple mappings). For library items, this will contain all mappings of all providers connected to the item.

The ProviderMapping has this structure:
```

Author: Applied in a7e38b2f3

- `item_id` (str): The external provider's item ID (e.g., `"spotify--track123"`)
- `provider_domain` (str): The external provider identifier (e.g., `"spotify"`, `"apple_music"`, `"tidal"`)
- `provider_instance` (str): Optional provider instance ID if multiple instances of same provider exist

> - `media_type` (MediaType): Type of media item (track, album, artist, playlist)

Reviewer: ```suggestion
- `item_id` (str): The music provider's item ID (e.g., `"spotify--track123"`)
- `provider_domain` (str): The musicprovider's domain (e.g., `"spotify"`, `"apple_music"`, `"tidal"`)
- `provider_instance` (str): The provider instance ID (to handle multiple instances of the same provider).
- some more details such as the quality (relevant in case of album or track)
```

Author: Applied in a7e38b2f3

**Important patterns:**

> - Library items are identified by `item.provider == "library"` where `item.item_id` is the library DB ID

Reviewer: ```suggestion
- A media item itself also has an item_id and provider attribute. For an item that comes straight from the musicprovider itself (so not from the MA library) the provider attribute will be set to the instance_id of that provider.
- Library items are identified by `item.provider == "library"` where `item.item_id` is the library DB ID
```

Author: Applied in a7e38b2f3

- `provider_domain='library'` **never exists** in provider_mappings — library items don't have mappings to themselves

> - To resolve a provider item to its library equivalent, use: `await mass.music.tracks.get_library_item_by_prov_id(item_id, provider_instance_id_or_domain)`

Reviewer: ```suggestion
- To resolve a provider item to its library equivalent, use: `await mass.music.<mediatype>.get_library_item_by_prov_id(item_id, provider_instance_id_or_domain)`
```

Author: Applied in a7e38b2f3

- **Logs:** `$HOME/.musicassistant/musicassistant.log` (current), `musicassistant.log.1`, `.log.2`, etc. for older rotated logs.
- **Database:** `$HOME/.musicassistant/library.db` — query via `sqlite3`. **Only execute SELECT queries** — never write to a live database.

### Provider Mappings

Every `MediaItem` (track, album, artist, playlist) has a `provider_mappings` attribute containing the exact mappings of the item on (each) provider. For a MediaItem that comes from a musicprovider directly, this will usually contain one single ID-mapping (although it is possible that a provider has multiple mappings). For library items, this will contain all mappings of all providers connected to the item.

The ProviderMapping has this structure:

- `item_id` (str): The music provider's item ID (e.g., `"spotify--track123"`)
- `provider_domain` (str): The musicprovider's domain (e.g., `"spotify"`, `"apple_music"`, `"tidal"`)
- `provider_instance` (str): The provider instance ID (to handle multiple instances of the same provider).

> - some more details such as the quality (relevant in case of album or track)

Author: This documentation describes the Python ProviderMapping model attribute, not the raw database schema. The ProviderMapping class maps the database's provider_item_id column to the item_id Python attribute, which is the correct interface developers interact with. Marcel's review confirmed this is the right level of abstraction.
